src/main/kotlin/se/zensum/katie/Main.kt

```diff
@@ -0,0 +1,128 @@
+package se.zensum.katie
+
+import franz.JobStatus
+import franz.ProducerBuilder
+import franz.WorkerBuilder
+import franz.producer.ProduceResult
+import khttp.responses.Response
+import kotlinx.coroutines.experimental.future.await
+import kotlinx.coroutines.experimental.runBlocking
+import mu.KotlinLogging
+import org.apache.kafka.common.errors.TimeoutException
+import org.jetbrains.ktor.http.HttpStatusCode
+import se.zensum.idempotenceconnector.IdempotenceStore
+import se.zensum.webhook.PayloadOuterClass
+import java.net.URL
+import java.util.concurrent.CompletableFuture
+
+
+fun getEnv(e: String, default: String? = null): String = System.getenv()[e] ?: default
+    ?: throw RuntimeException("Missing environment variable $e and no default value is given.")
+
+private val logger = KotlinLogging.logger("main")
+private val routes: Map<String, Route> = readRoutes()
+private val KAFKA_HOST: String = getEnv("KAFKA_HOST", "kafka")
+private val idempotenceStore = IdempotenceStore(set = "katie")
+private val producer = ProducerBuilder.ofByteArray.option("bootstrap.servers", KAFKA_HOST)
+    .create()
+
+fun main(args: Array<String>) = runBlocking {
+    WorkerBuilder.ofByteArray
+        .subscribedTo(routes.keys)
+        .groupId("katie")
+        .handlePiped {
+            val topic: String = it.value?.topic() ?: return@handlePiped JobStatus.PermanentFailure
+            val offset: Long = it.value?.offset() ?: return@handlePiped JobStatus.PermanentFailure
+            val url: URL? = routes[topic]!!.url
+            val responseTopic: String = routes[topic]!!.responseTopic
+            it
+                .require("URL is not null for topic ($topic)") { url != null }
+                .map { Payload(it.value()) }
+                .advanceIf("Not in idempotence store") { checkIdempotenceStore(it, offset) }
+                .execute("Send to $url") { send(url!!, it, responseTopic) }
+                .execute("Write in idempotence store") { writeToIdempotenceStore(it, offset) }
+                .end()
+        }.start()
+}
+
+private suspend fun writeToIdempotenceStore(it: PayloadOuterClass.Payload, offset: Long): Boolean {
+    return when (it.repeating) {
+        true -> idempotenceStore.put(it.flakeId.toString() + offset)
```

Comment:
Consider setting an additional character between `flakeId` and `offset` (like `:` or `-`) to avoid a collision in the extremely unlikely event of
`15178244361984|173972351|1` (flakeId) + `1` (offset)
colliding with
`15178244361984|173972351|11` (flakeId) + (no offset since repeated is false)
since both would result in `1517824436198417397235111`.

This would require ten request to be generated on the same node within a single millisecond, and the offset adding up "perfectly", so it is *very* unlikely too happen. If making these change, caution should be taken so that the delimiter does not affect the URL safety of the generated string.